synapse/handlers/oidc_handler.py

```diff
@@ -885,10 +890,12 @@ async def _map_userinfo_to_user(
             "Retrieved user attributes from user mapping provider: %r", attributes
         )
 
-        if not attributes["localpart"]:
-            raise MappingException("localpart is empty")
-
-        localpart = map_username_to_mxid_localpart(attributes["localpart"])
```

Comment:
This got pushed into the mapping provider, which matches SAML and gives mapping providers a bit more control (e.g. the dot-replace vs. hex-encode options that SAML has).

Reply:
Is that a change in the mapping provider API?

Reply:
The SAML mapping provider (implicitly) must provide a valid localpart. I think this was kind of the assumption with the OIDC one as well, but I'm not sure. I failed to update the documentation to make this explicit for both of them, but I meant to! 😄

Reply:
I pushed some documentation improvements (and added something to `UPGRADE.rst`). I suspect not many people have their own mapping provider, but I'm not sure.